### PR TITLE
[kube-prometheus-stack] allow setting annotations for admission webhooks

### LINKED ISSUE #4686: [kube-prometheus-stack] Allow custom annotations to be added to admission webhook configurations
### Is your feature request related to a problem ?

The admission webhook patch job for Prometheus Operator fails when deployed in ArgoCD's `PreSync` phase (as [suggested in the values file](https://github.com/prometheus-community/helm-charts/blob/2a72f34fa1d5d3d7dc5ee1efbc4c3770faacbdb5/charts/kube-prometheus-stack/values.yaml#L2473)), because it requires the `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` to be present.

Currently these resources do not support custom annotations, so there is no way to ensure they are created in the same phase as the patch job. 

### Describe the solution you'd like.

Allow custom annotations to be added to these webhook configurations and any other resources required by the patch job.

### Describe alternatives you've considered.

* Attempting to get everything ordered correctly in the `Sync` phase using sync waves. This creates some different issues relating to startup order/race conditions between the operator and patch job.

### Additional context.

_No response_

### PR DESCRIPTION
**Pull request by DrFaust92**
#### Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- fixes #4686

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

File: charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/validatingWebhookConfiguration.yaml [hunk 1]
```diff
@@ -3,14 +3,14 @@ apiVersion: admissionregistration.k8s.io/v1
...
 metadata:
   name:  {{ template "kube-prometheus-stack.fullname" . }}-admission
-{{- if .Values.prometheusOperator.admissionWebhooks.certManager.enabled }}
   annotations:
-    certmanager.k8s.io/inject-ca-from: {{ printf "%s/%s-admission" (include "kube-prometheus-stack.namespace" .) (include "kube-prometheus-stack.fullname" .) | quote }}
-    cert-manager.io/inject-ca-from: {{ printf "%s/%s-admission" (include "kube-prometheus-stack.namespace" .) (include "kube-prometheus-stack.fullname" .) | quote }}
-{{- end }}
+    {{- include "kube-prometheus-stack.prometheus-operator-webhook.annotations" $ | trim |nindent 4 }}
+    {{- with .Values.prometheusOperator.admissionWebhooks.validatingWebhookConfiguration.annotations }}
+    {{ toYaml . }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/validatingWebhookConfiguration.yaml:9]: this sound suspicious to me. no indents?

Could you please add `prometheusOperator.admissionWebhooks.validatingWebhookConfiguration.annotations` to an existing ci test values file?

**Review by jkroepke — CHANGES_REQUESTED** [review]

**Comment by DrFaust92** [comment]
Yep, will add.

File: charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/validatingWebhookConfiguration.yaml (hunk 1 above)
Comment (DrFaust92) [reply charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/validatingWebhookConfiguration.yaml:9]: done

File: charts/kube-prometheus-stack/ci/04-prometheus-operator-webhook-values.yaml [hunk 2]
```diff
@@ -1,4 +1,7 @@
 prometheusOperator:
   admissionWebhooks:
+    validatingWebhookConfiguration:
+      annotations:
+        test: test1
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/ci/04-prometheus-operator-webhook-values.yaml:5]:
```suggestion
        test: test1
        test2: test3
```

**Comment by DrFaust92** [comment]
thanks @jkroepke i think i fixed it!

File: charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/mutatingWebhookConfiguration.yaml [hunk 3]
```diff
@@ -3,14 +3,14 @@ apiVersion: admissionregistration.k8s.io/v1
...
 metadata:
   name:  {{ template "kube-prometheus-stack.fullname" . }}-admission
-{{- if .Values.prometheusOperator.admissionWebhooks.certManager.enabled }}
   annotations:
-    certmanager.k8s.io/inject-ca-from: {{ printf "%s/%s-admission" (include "kube-prometheus-stack.namespace" .) (include "kube-prometheus-stack.fullname" .) | quote }}
-    cert-manager.io/inject-ca-from: {{ printf "%s/%s-admission" (include "kube-prometheus-stack.namespace" .) (include "kube-prometheus-stack.fullname" .) | quote }}
-{{- end }}
+    {{- include "kube-prometheus-stack.prometheus-operator-webhook.annotations" $ | trim |nindent 4 }}
+    {{- with .Values.prometheusOperator.admissionWebhooks.mutatingWebhookConfiguration.annotations }}
+    {{ toYaml . | trim | nindent 4}}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/mutatingWebhookConfiguration.yaml:9]:
```suggestion
    {{- toYaml . | nindent 4}}
```

File: charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/mutatingWebhookConfiguration.yaml [hunk 4]
```diff
@@ -3,14 +3,14 @@ apiVersion: admissionregistration.k8s.io/v1
...
+    {{- end }}
   labels:
     app: {{ template "kube-prometheus-stack.name" $ }}-admission
-    {{- include "kube-prometheus-stack.prometheus-operator-webhook.labels" $ | nindent 4 }}
+    {{- include "kube-prometheus-stack.prometheus-operator-webhook.labels" $ | trim | nindent 4 }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/mutatingWebhookConfiguration.yaml:13]:
```suggestion
    {{- include "kube-prometheus-stack.prometheus-operator-webhook.labels" $ | nindent 4 }}
```

File: charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/validatingWebhookConfiguration.yaml [hunk 5]
```diff
@@ -3,14 +3,14 @@ apiVersion: admissionregistration.k8s.io/v1
...
 metadata:
   name:  {{ template "kube-prometheus-stack.fullname" . }}-admission
-{{- if .Values.prometheusOperator.admissionWebhooks.certManager.enabled }}
   annotations:
-    certmanager.k8s.io/inject-ca-from: {{ printf "%s/%s-admission" (include "kube-prometheus-stack.namespace" .) (include "kube-prometheus-stack.fullname" .) | quote }}
-    cert-manager.io/inject-ca-from: {{ printf "%s/%s-admission" (include "kube-prometheus-stack.namespace" .) (include "kube-prometheus-stack.fullname" .) | quote }}
-{{- end }}
+    {{- include "kube-prometheus-stack.prometheus-operator-webhook.annotations" $ | trim | nindent 4 }}
+    {{- with .Values.prometheusOperator.admissionWebhooks.validatingWebhookConfiguration.annotations }}
+    {{ toYaml . | trim | nindent 4}}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/validatingWebhookConfiguration.yaml:9]:
```suggestion
    {{- toYaml . | nindent 4}}
```

File: charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/validatingWebhookConfiguration.yaml [hunk 6]
```diff
@@ -3,14 +3,14 @@ apiVersion: admissionregistration.k8s.io/v1
...
+    {{- end }}
   labels:
     app: {{ template "kube-prometheus-stack.name" $ }}-admission
-    {{- include "kube-prometheus-stack.prometheus-operator-webhook.labels" $ | nindent 4 }}
+    {{- include "kube-prometheus-stack.prometheus-operator-webhook.labels" $ | trim |nindent 4 }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/prometheus-operator/admission-webhooks/validatingWebhookConfiguration.yaml:13]:
```suggestion
    {{- include "kube-prometheus-stack.prometheus-operator-webhook.labels" $ | nindent 4 }}
```

File: charts/kube-prometheus-stack/values.yaml [hunk 7]
```diff
@@ -2238,6 +2238,13 @@ prometheusOperator:
     namespaceSelector: {}
     objectSelector: {}
 
+    mutatingWebhookConfiguration:
+      annotations: {}
+      #   argocd.argoproj.io/hook: PreSync
+
+    validatingWebhookConfiguration:
+      annotations:
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/values.yaml:2246]:
```suggestion
      annotations: {}
```

**Review by jkroepke — COMMENTED** [review]
We are using {{- toyaml | nindent }}

everywhere, that should work, too. the hypen at start trims, too.

**Review by jkroepke — APPROVED** [review]

**Comment by DrFaust92** [comment]
Thanks @jkroepke , im thinking of a follow up, i see that `admissionWebhooks.annotations` is not propagated to all resources and would like to add it as well

**Pull request merged**